File: graze/graze-custom-node-manager/web/static/js/spam_blocker.js
// Spam Blocker Manager
import { api } from './api.js';

const CATEGORIES = [
    { id: 'cryptobros', name: 'Cryptobros', description: 'Crypto scammers and trading signal spammers' },
    { id: 'engagement_farmers', name: 'Engagement Farmers', description: 'Follow-for-follow and engagement bait accounts' },
    // { id: 'thirst_traps', name: 'Thirst Traps', description: 'Adult content promoters and OnlyFans spam' },  // DISABLED: Needs refinement
    { id: 'huns', name: 'MLM Huns', description: 'MLM/pyramid scheme promoters and boss babes' },
    { id: 'manosphere_grifters', name: 'Manosphere Grifters', description: 'Alpha male coaches, course sellers, and toxic masculinity hustlers' },
    { id: 'dropshitters', name: 'Dropshitters', description: 'Ecommerce spam and wholesale DM merchants' },
    { id: 'prize_pigs', name: 'Prize Pigs', description: 'Fake giveaway and "you won!" scammers' },
    { id: 'wolf_of_wall_street_wannabes', name: 'Wolf of Wall Street Wannabes', description: 'Forex/stock trading signal scammers and day trading gurus' },
    { id: 'hustle_porn', name: 'Hustle Porn', description: 'Get rich quick and passive income fantasies' },
    { id: 'link_leeches', name: 'Link Leeches', description: 'Generic "link in bio" and "DM me" spammers' }
];

let currentCategory = null;
let currentTab = 'bio_terms';
let categoryData = {};

export async function showManager() {
    const modal = document.getElementById('spam-blocker-modal');
    modal.style.display = 'flex';
    
    await loadCategories();
    renderCategoryList();
}

async function loadCategories() {
    for (const cat of CATEGORIES) {
        const data = await api.get(`/api/spam-blocker/${cat.id}`);
        categoryData[cat.id] = data;
    }
}

function renderCategoryList() {
    const list = document.getElementById('spam-categories-list');
    list.innerHTML = CATEGORIES.map(cat => `
        <div class="category-item ${currentCategory === cat.id ? 'active' : ''}" 
             onclick="window.spamBlockerModule.selectCategory('${cat.id}')" 
             style="padding: 0.75rem; margin-bottom: 0.5rem; border: 1px solid #e5e5e5; border-radius: 4px; cursor: pointer; ${currentCategory === cat.id ? 'background: #f0f9ff; border-color: #3b82f6;' : ''}">
            <strong>${cat.name}</strong>
            <div style="font-size: 0.85rem; color: #666; margin-top: 0.25rem;">${cat.description}</div>
        </div>
    `).join('');
}

function selectCategory(categoryId) {
    currentCategory = categoryId;
    currentTab = 'bio_terms';
    renderCategoryList();
    renderEditor();
}

function renderEditor() {
    if (!currentCategory) return;
    
    const editor = document.getElementById('spam-editor');
    const category = CATEGORIES.find(c => c.id === currentCategory);
    
    editor.style.display = 'block';
    document.getElementById('spam-category-name').textContent = category.name;
    
    showSpamTab('bio_terms');
}

function showSpamTab(tab) {
    currentTab = tab;
    
    document.querySelectorAll('.spam-tab-btn').forEach(btn => {
        btn.classList.remove('active');
        btn.style.borderBottom = '3px solid transparent';
    });
    
    document.querySelectorAll('.spam-tab-content').forEach(content => {
        content.style.display = 'none';
    });
    
    const activeBtn = document.querySelector(`.spam-tab-btn[onclick*="${tab}"]`);
    if (activeBtn) {
        activeBtn.classList.add('active');
        activeBtn.style.borderBottom = '3px solid #6366f1';
    }
    
    document.getElementById(`spam-${tab}-tab`).style.display = 'block';
    renderTermsList(tab);
}

function renderTermsList(tab) {
    const data = categoryData[currentCategory];
    const terms = data[tab] || [];
    const list = document.getElementById(`spam-${tab}-list`);
    
    const isExceptions = tab === 'bio_exceptions';
    const label = isExceptions ? 'Exception' : 'Term';
    
    list.innerHTML = terms.map(term => `
        <div style="display: flex; justify-content: space-between; align-items: center; padding: 0.5rem; border: 1px solid #e5e5e5; border-radius: 4px; margin-bottom: 0.25rem; ${isExceptions ? 'background: #f0fdf4;' : ''}">
            <span>${term}</span>
            <button onclick="window.spamBlockerModule.removeTerm('${tab}', '${term.replace(/'/g, "\\'")}')">×</button>
        </div>
    `).join('');
}

function addTerm(tab) {
    const input = document.getElementById(`spam-${tab}-input`);
    const term = input.value.trim();
    
    if (!term) return;
    
    if (!categoryData[currentCategory][tab]) {
        categoryData[currentCategory][tab] = [];
    }
    
    if (!categoryData[currentCategory][tab].includes(term)) {
        categoryData[currentCategory][tab].push(term);
        renderTermsList(tab);
    }
    
    input.value = '';
}

function removeTerm(tab, term) {
    const data = categoryData[currentCategory];
    data[tab] = data[tab].filter(t => t !== term);
    renderTermsList(tab);
}

async function saveCategory() {
    await api.post(`/api/spam-blocker/${currentCategory}`, categoryData[currentCategory]);
    alert('Category saved!');
}

function closeManager() {
    document.getElementById('spam-blocker-modal').style.display = 'none';
}

window.spamBlockerModule = {
    showManager,
    selectCategory,
    showSpamTab,
    addTerm,
    removeTerm,
    saveCategory,
    closeManager
};

File: graze/graze-custom-node-manager/nodes/spam_blocker.py
"""
Spam Blocker - Block spam accounts and content based on bio patterns and post content
"""
import json
import os

metadata = {
    "id": "spam_blocker",
    "name": "Spam Blocker",
    "description": "Block spam accounts by scanning user bios and post content. Targets cryptobros, engagement farmers, adult content promoters, MLM schemes, and grifters.",
    "color": "yellow",
    "version": "1.0.0",
    "author": "Custom Nodes System",
    "tags": ["spam", "moderation", "bio", "scam"],
    "configurable": {},
    "manageable": True
}

def load_category(category):
    """Load spam category data"""
    data_path = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "data", "spam_blocker", f"{category}.json")
    try:
        with open(data_path, "r") as f:
            return json.load(f)
    except:
        return {}

def build_bio_filter(terms, exceptions=None):
    """Build filter for bio scanning with exception support"""
    if not terms:
        return None
    
    if exceptions:
        # Allow if bio has exception terms, otherwise block if has spam terms
        return {
            "or": [
                {"regex_any": ["hydrated_metadata.user.description", exceptions, True, False]},  # has exceptions = allow
                {"regex_none": ["hydrated_metadata.user.description", terms, True, False]}  # no spam terms = allow
            ]
        }
    
    return {"regex_none": ["hydrated_metadata.user.description", terms, True, False]}

def build_post_filter(terms):
    """Build filter for post content"""
    if not terms:
        return None
    fields = ["text", "embed.external.title", "embed.external.description", "embed.alt", "embed.images[*].alt"]
    return {"or": [{"regex_none": [field, terms, True, False]} for field in fields]}

def build_hashtag_filter(hashtags):
    """Build filter for hashtags"""
    if not hashtags:
        return None
    formatted = [f"#{tag}" for tag in hashtags]
    return {"or": [
        {"regex_none": ["text", formatted, True, False]}
    ]}

def build_category_filter(category_name, category_data, toggle_prefix):
    """Build complete filter for a spam category with separate toggles"""
    checks = []
    
    bio_terms = category_data.get("bio_terms", [])
    bio_exceptions = category_data.get("bio_exceptions", [])
    post_terms = category_data.get("post_terms", [])
    hashtags = category_data.get("hashtags", [])
    
    if bio_terms:
        bio_filter = build_bio_filter(bio_terms, bio_exceptions if bio_exceptions else None)
        if bio_filter:
            checks.append({
                "or": [
                    {"param_compare": [f"${toggle_prefix}_BIO", "==", False]},
                    bio_filter
                ]
            })
    
    if post_terms:
        post_filter = build_post_filter(post_terms)
        if post_filter:
            checks.append({
                "or": [
                    {"param_compare": [f"${toggle_prefix}_POSTS", "==", False]},
                    post_filter
                ]
            })
    
    if hashtags:
        hashtag_filter = build_hashtag_filter(hashtags)
        if hashtag_filter:
            checks.append({
                "or": [
                    {"param_compare": [f"${toggle_prefix}_HASHTAGS", "==", False]},
                    hashtag_filter
                ]
            })
    
    if not checks:
        return None
    
    return {"and": checks}

def get_manifest(options=None):
    """Generate the filter manifest"""
    
    categories = {
        "BLOCK_CRYPTOBROS": ("cryptobros", "Cryptobros", "Crypto scammers and trading signal spammers"),
        "BLOCK_ENGAGEMENT_FARMERS": ("engagement_farmers", "Engagement Farmers", "Follow-for-follow and engagement bait accounts"),
        # "BLOCK_THIRST_TRAPS": ("thirst_traps", "Thirst Traps", "Adult content promoters and OnlyFans spam"),  # DISABLED
        "BLOCK_HUNS": ("huns", "MLM Huns", "MLM/pyramid scheme promoters and boss babes"),
        "BLOCK_MANOSPHERE_GRIFTERS": ("manosphere_grifters", "Manosphere Grifters", "Alpha male coaches, course sellers, and toxic masculinity hustlers"),
        "BLOCK_DROPSHITTERS": ("dropshitters", "Dropshitters", "Ecommerce spam and wholesale DM merchants"),
        "BLOCK_PRIZE_PIGS": ("prize_pigs", "Prize Pigs", "Fake giveaway and 'you won!' scammers"),
        "BLOCK_WOLF_OF_WALL_STREET_WANNABES": ("wolf_of_wall_street_wannabes", "Wolf of Wall Street Wannabes", "Forex/stock trading signal scammers and day trading gurus"),
        "BLOCK_HUSTLE_PORN": ("hustle_porn", "Hustle Porn", "Get rich quick and passive income fantasies"),
        "BLOCK_LINK_LEECHES": ("link_leeches", "Link Leeches", "Generic 'link in bio' and 'DM me' spammers")
    }
    
    filters = []
    params = [
        {"name": "WHITELIST_SPAM_BLOCKER", "type": "list", "description": "Exempt users from spam blocker restrictions.", "displayName": "Whitelist", "exampleValue": []}
    ]
    groups = []
    
    for toggle_prefix, (file_name, group_name, group_desc) in categories.items():
        category_data = load_category(file_name)
        category_filter = build_category_filter(file_name, category_data, toggle_prefix)
        
        if category_filter:
            filters.append(category_filter)
            
            # Add parameter group
            group_id = file_name
            groups.append({
                "id": group_id,
                "name": group_name,
                "description": group_desc
            })
            
            # Add toggles for bio, posts, hashtags
            if category_data.get("bio_terms"):
                params.append({
                    "name": f"{toggle_prefix}_BIO",
                    "type": "toggle",
                    "description": "Scan user bios for spam patterns",
                    "displayName": f"Block {group_name} (Bio)",
                    "exampleValue": False,
                    "group": group_id
                })
            
            if category_data.get("post_terms"):
                params.append({
                    "name": f"{toggle_prefix}_POSTS",
                    "type": "toggle",
                    "description": "Scan post content for spam patterns",
                    "displayName": f"Block {group_name} (Posts)",
                    "exampleValue": False,
                    "group": group_id
                })
            
            if category_data.get("hashtags"):
                params.append({
                    "name": f"{toggle_prefix}_HASHTAGS",
                    "type": "toggle",
                    "description": "Scan hashtags for spam patterns",
                    "displayName": f"Block {group_name} (Hashtags)",
                    "exampleValue": False,
                    "group": group_id
                })
    
    # Wrap all filters with whitelist check
    final_filter = {
        "or": [
            {"each": ["$WHITELIST_SPAM_BLOCKER", {"list_member": ["$WHITELIST_SPAM_BLOCKER_ITEM", "in"]}]},
            {"and": filters}
        ]
    }
    
    return {
        "filter": {
            "and": [final_filter],
            "metadata": {
                "color": "yellow",
                "customNodeParameters": params,
                "customNodeParameterGroups": groups
            }
        }
    }
